hogql_parser/parser.cpp

```diff
@@ -0,0 +1,1367 @@
+    PyObject* order_by = visitAsPyObjectOrNone(ctx->winOrderByClause());
+    PyObject* frame_method = frame_ctx && frame_ctx->RANGE()  ? PyUnicode_FromString("RANGE")
+                             : frame_ctx && frame_ctx->ROWS() ? PyUnicode_FromString("ROWS")
+                                                              : Py_NewRef(Py_None);
+    PyObject* frame_start = PyTuple_Check(frame) ? PyTuple_GetItem(frame, 0) : frame;
+    PyObject* frame_end = PyTuple_Check(frame) ? PyTuple_GetItem(frame, 1) : Py_NewRef(Py_None);
+    return build_ast_node(
+        "WindowExpr", "{s:N,s:N,s:N,s:N,s:N}", "partition_by", partition_by, "order_by", order_by, "frame_method",
+        frame_method, "frame_start", frame_start, "frame_end", frame_end
+    );
+  }
+
+  VISIT(WinPartitionByClause) { return visit(ctx->columnExprList()); }
+
+  VISIT(WinOrderByClause) { return visit(ctx->orderExprList()); }
+
+  VISIT(WinFrameClause) { return visit(ctx->winFrameExtend()); }
+
+  VISIT(FrameStart) { return visit(ctx->winFrameBound()); }
+
+  VISIT(FrameBetween) {
+    return Py_BuildValue("NN", visitAsPyObject(ctx->winFrameBound(0)), visitAsPyObject(ctx->winFrameBound(1)));
+  }
+
+  VISIT(WinFrameBound) {
+    if (ctx->PRECEDING() || ctx->FOLLOWING()) {
+      PyObject* number;
+      if (ctx->numberLiteral()) {
+        number = PyObject_GetAttrString(visitAsPyObject(ctx->numberLiteral()), "value");
+      } else {
+        number = Py_NewRef(Py_None);
+      }
+      return build_ast_node(
+          "WindowFrameExpr", "{s:s,s:N}", "frame_type", ctx->PRECEDING() ? "PRECEDING" : "FOLLOWING", "frame_value",
+          number
+      );
+    } else {
+      return build_ast_node("WindowFrameExpr", "{s:s}", "frame_type", "CURRENT ROW");
+    }
+  }
+
+  VISIT(Expr) { return visit(ctx->columnExpr()); }
+
+  VISIT_UNSUPPORTED(ColumnTypeExprSimple)
+
+  VISIT_UNSUPPORTED(ColumnTypeExprNested)
+
+  VISIT_UNSUPPORTED(ColumnTypeExprEnum)
+
+  VISIT_UNSUPPORTED(ColumnTypeExprComplex)
+
+  VISIT_UNSUPPORTED(ColumnTypeExprParam)
+
+  VISIT(ColumnExprList) { return visitPyListOfObjects(ctx->columnExpr()); }
+
+  VISIT(ColumnExprTernaryOp) {
+    return build_ast_node(
+        "Call", "{s:s, s:[O,O,O]}", "name", "if", "args", visitAsPyObject(ctx->columnExpr(0)),
+        visitAsPyObject(ctx->columnExpr(1)), visitAsPyObject(ctx->columnExpr(2))
+    );
+  }
+
+  VISIT(ColumnExprAlias) {
+    string alias;
+    if (ctx->alias()) {
+      alias = visitAsString(ctx->alias());
+    } else if (ctx->identifier()) {
+      alias = visitAsString(ctx->identifier());
+    } else if (ctx->STRING_LITERAL()) {
+      alias = unquote_string_terminal(ctx->STRING_LITERAL());
+    } else {
+      throw HogQLParsingException("A ColumnExprAlias must have the alias in some form");
+    }
+    PyObject* expr = visitAsPyObject(ctx->columnExpr());
+
+    if (find(RESERVED_KEYWORDS.begin(), RESERVED_KEYWORDS.end(), boost::algorithm::to_lower_copy(alias)) !=
```

Comment:
Are those `boost::algorithm::to_lower_copy(alias)` strings also getting cleared up, or should we explicitly `free` them?

Reply:
We don't use `new` anywhere, so no need to `delete` anything – this is on the stack